File: app/src/main/java/app/kamy/saatApp/domain/faraidh/FaraidhGraphBuilder.kt
package app.kamy.saatApp.domain.faraidh

import java.math.BigDecimal

object FaraidhGraphBuilder {

    fun build(
        profile: DeceasedProfile,
        input: HeirInput,
        activeShares: List<HeirShare>,
        blockedHeirs: List<BlockedHeir>
    ): FaraidhFamilyGraph {
        val nodes = mutableMapOf<String, FaraidhGraphNode>()
        val edges = mutableListOf<FaraidhGraphEdge>()

        val rootId = "deceased_root"
        val deceasedName = profile.name.ifBlank { "Deceased" }
        nodes[rootId] = FaraidhGraphNode(
            id = rootId,
            displayName = deceasedName,
            generationLevel = 0,
            relationType = null,
            status = GraphNodeStatus.ACTIVE,
            visualColorHex = "#1E293B"
        )

        // 1. Spouses (Generation 0)
        if (input.husbandCount > 0) {
            val id = "husband_1"
            val share = activeShares.find { it.type == HeirType.HUSBAND }
            nodes[id] = FaraidhGraphNode(
                id = id,
                displayName = "Husband",
                generationLevel = 0,
                relationType = HeirType.HUSBAND,
                status = if (share != null) GraphNodeStatus.ACTIVE else GraphNodeStatus.MAHJUB_HIRMAN,
                baseShareFraction = share?.fraction?.toDisplayString(),
                finalPercentage = share?.percentage?.toDouble() ?: 0.0,
                cashValue = share?.cashAmount ?: BigDecimal.ZERO,
                visualColorHex = if (share != null) "#10B981" else "#EF4444"
            )
            edges.add(FaraidhGraphEdge(rootId, id, GraphEdgeType.MARRIAGE))
        }

        for (i in 1..input.wifeCount) {
            val id = "wife_$i"
            val share = activeShares.find { it.type == HeirType.WIFE }
            nodes[id] = FaraidhGraphNode(
                id = id,
                displayName = "Wife $i",
                generationLevel = 0,
                relationType = HeirType.WIFE,
                status = if (share != null) GraphNodeStatus.ACTIVE else GraphNodeStatus.MAHJUB_HIRMAN,
                baseShareFraction = share?.fraction?.toDisplayString(),
                finalPercentage = (share?.percentage?.toDouble() ?: 0.0) / input.wifeCount,
                cashValue = (share?.cashAmount ?: BigDecimal.ZERO).divide(BigDecimal(input.wifeCount), 2, java.math.RoundingMode.HALF_UP),
                visualColorHex = if (share != null) "#10B981" else "#EF4444"
            )
            edges.add(FaraidhGraphEdge(rootId, id, GraphEdgeType.MARRIAGE))
        }

        // 2. Parents (Generation +1)
        if (input.fatherCount > 0) {
            val id = "father_1"
            val share = activeShares.find { it.type == HeirType.FATHER }
            val blocked = blockedHeirs.find { it.type == HeirType.FATHER }
            val status = when {
                blocked?.reason == BlockingReasonKey.OUT_OF_WEDLOCK -> GraphNodeStatus.EXCLUDED_BY_LAW
                blocked != null -> GraphNodeStatus.MAHJUB_HIRMAN
                else -> GraphNodeStatus.ACTIVE
            }
            nodes[id] = FaraidhGraphNode(
                id = id,
                displayName = "Father",
                generationLevel = 1,
                relationType = HeirType.FATHER,
                status = status,
                baseShareFraction = share?.fraction?.toDisplayString(),
                finalPercentage = share?.percentage?.toDouble() ?: 0.0,
                cashValue = share?.cashAmount ?: BigDecimal.ZERO,
                disqualificationReasonId = blocked?.reason?.name?.lowercase(),
                visualColorHex = if (status == GraphNodeStatus.ACTIVE) "#10B981" else "#EF4444"
            )
            edges.add(FaraidhGraphEdge(id, rootId, GraphEdgeType.PARENT_CHILD))
        }

        if (input.motherCount > 0) {
            val id = "mother_1"
            val share = activeShares.find { it.type == HeirType.MOTHER }
            nodes[id] = FaraidhGraphNode(
                id = id,
                displayName = "Mother",
                generationLevel = 1,
                relationType = HeirType.MOTHER,
                status = if (share != null) GraphNodeStatus.ACTIVE else GraphNodeStatus.MAHJUB_HIRMAN,
                baseShareFraction = share?.fraction?.toDisplayString(),
                finalPercentage = share?.percentage?.toDouble() ?: 0.0,
                cashValue = share?.cashAmount ?: BigDecimal.ZERO,
                visualColorHex = if (share != null) "#10B981" else "#EF4444"
            )
            edges.add(FaraidhGraphEdge(id, rootId, GraphEdgeType.PARENT_CHILD))
        }

        // 3. Children (Generation -1)
        for (i in 1..input.sonCount) {
            val id = "son_$i"
            val share = activeShares.find { it.type == HeirType.SON }
            nodes[id] = FaraidhGraphNode(
                id = id,
                displayName = "Son $i",
                generationLevel = -1,
                relationType = HeirType.SON,
                status = if (share != null) GraphNodeStatus.ACTIVE else GraphNodeStatus.MAHJUB_HIRMAN,
                baseShareFraction = share?.fraction?.toDisplayString(),
                finalPercentage = (share?.percentage?.toDouble() ?: 0.0) / input.sonCount,
                cashValue = (share?.cashAmount ?: BigDecimal.ZERO).divide(BigDecimal(input.sonCount), 2, java.math.RoundingMode.HALF_UP),
                visualColorHex = if (share != null) "#10B981" else "#EF4444"
            )
            edges.add(FaraidhGraphEdge(rootId, id, GraphEdgeType.PARENT_CHILD))
        }

        for (i in 1..input.daughterCount) {
            val id = "daughter_$i"
            val share = activeShares.find { it.type == HeirType.DAUGHTER }
            nodes[id] = FaraidhGraphNode(
                id = id,
                displayName = "Daughter $i",
                generationLevel = -1,
                relationType = HeirType.DAUGHTER,
                status = if (share != null) GraphNodeStatus.ACTIVE else GraphNodeStatus.MAHJUB_HIRMAN,
                baseShareFraction = share?.fraction?.toDisplayString(),
                finalPercentage = (share?.percentage?.toDouble() ?: 0.0) / input.daughterCount,
                cashValue = (share?.cashAmount ?: BigDecimal.ZERO).divide(BigDecimal(input.daughterCount), 2, java.math.RoundingMode.HALF_UP),
                visualColorHex = if (share != null) "#10B981" else "#EF4444"
            )
            edges.add(FaraidhGraphEdge(rootId, id, GraphEdgeType.PARENT_CHILD))
        }

        // 4. Disqualified heirs
        for ((idx, dq) in input.disqualifiedHeirs.withIndex()) {
            val id = "dq_${dq.type.name.lowercase()}_$idx"
            nodes[id] = FaraidhGraphNode(
                id = id,
                displayName = "${dq.type.name.lowercase().replaceFirstChar { it.titlecase() }} (Excluded)",
                generationLevel = if (dq.type == HeirType.SON || dq.type == HeirType.DAUGHTER) -1 else 0,
                relationType = dq.type,
                status = GraphNodeStatus.EXCLUDED_BY_LAW,
                disqualificationReasonId = dq.reason.name.lowercase(),
                visualColorHex = "#EF4444"
            )
            edges.add(FaraidhGraphEdge(rootId, id, GraphEdgeType.PARENT_CHILD))
        }

        return FaraidhFamilyGraph(
            rootDeceasedId = rootId,
            nodes = nodes,
            edges = edges
        )
    }
}

File: iosApp/Saat/Domain/Models/FaraidhGraphBuilder.swift
import Foundation

struct FaraidhGraphBuilder {
    static func build(
        profile: DeceasedProfile,
        input: HeirInput,
        activeShares: [HeirShare],
        blockedHeirs: [BlockedHeir]
    ) -> FaraidhFamilyGraph {
        var nodes: [String: FaraidhGraphNode] = [:]
        var edges: [FaraidhGraphEdge] = []
        
        let rootId = "deceased_root"
        let deceasedName = profile.name.isEmpty ? "Deceased" : profile.name
        nodes[rootId] = FaraidhGraphNode(
            id: rootId,
            displayName: deceasedName,
            generationLevel: 0,
            relationType: nil,
            status: .active,
            visualColorHex: "#1E293B"
        )
        
        // 1. Spouses (Generation 0)
        if input.husbandCount > 0 {
            let id = "husband_1"
            let share = activeShares.first { $0.type == .husband }
            nodes[id] = FaraidhGraphNode(
                id: id,
                displayName: "Husband",
                generationLevel: 0,
                relationType: .husband,
                status: share != nil ? .active : .mahjubHirman,
                baseShareFraction: share?.fraction.toDisplayString(),
                finalPercentage: Double(truncating: (share?.percentage ?? .zero) as NSNumber),
                cashValue: share?.cashAmount ?? .zero,
                visualColorHex: share != nil ? "#10B981" : "#EF4444"
            )
            edges.append(FaraidhGraphEdge(fromId: rootId, toId: id, type: .marriage))
        }
        
        if input.wifeCount > 0 {
            for i in 1...input.wifeCount {
                let id = "wife_\(i)"
                let share = activeShares.first { $0.type == .wife }
                let wifePct = Double(truncating: ((share?.percentage ?? .zero) / Decimal(input.wifeCount)) as NSNumber)
                let wifeCash = (share?.cashAmount ?? .zero) / Decimal(input.wifeCount)
                nodes[id] = FaraidhGraphNode(
                    id: id,
                    displayName: "Wife \(i)",
                    generationLevel: 0,
                    relationType: .wife,
                    status: share != nil ? .active : .mahjubHirman,
                    baseShareFraction: share?.fraction.toDisplayString(),
                    finalPercentage: wifePct,
                    cashValue: wifeCash,
                    visualColorHex: share != nil ? "#10B981" : "#EF4444"
                )
                edges.append(FaraidhGraphEdge(fromId: rootId, toId: id, type: .marriage))
            }
        }
        
        // 2. Parents (Generation +1)
        if input.fatherCount > 0 {
            let id = "father_1"
            let share = activeShares.first { $0.type == .father }
            let blocked = blockedHeirs.first { $0.type == .father }
            let status: GraphNodeStatus = {
                if blocked?.reason == .outOfWedlock { return .excludedByLaw }
                if blocked != nil { return .mahjubHirman }
                return .active
            }()
            nodes[id] = FaraidhGraphNode(
                id: id,
                displayName: "Father",
                generationLevel: 1,
                relationType: .father,
                status: status,
                baseShareFraction: share?.fraction.toDisplayString(),
                finalPercentage: Double(truncating: (share?.percentage ?? .zero) as NSNumber),
                cashValue: share?.cashAmount ?? .zero,
                disqualificationReasonId: blocked?.reason.rawValue.lowercased(),
                visualColorHex: status == .active ? "#10B981" : "#EF4444"
            )
            edges.append(FaraidhGraphEdge(fromId: id, toId: rootId, type: .parentChild))
        }
        
        if input.grandfatherCount > 0 {
            let id = "grandfather_1"
            let share = activeShares.first { $0.type == .grandfather }
            let blocked = blockedHeirs.first { $0.type == .grandfather }
            let status: GraphNodeStatus = {
                if blocked?.reason == .outOfWedlock { return .excludedByLaw }
                if blocked != nil { return .mahjubHirman }
                return .active
            }()
            nodes[id] = FaraidhGraphNode(
                id: id,
                displayName: "Grandfather",
                generationLevel: 2,
                relationType: .grandfather,
                status: status,
                baseShareFraction: share?.fraction.toDisplayString(),
                finalPercentage: Double(truncating: (share?.percentage ?? .zero) as NSNumber),
                cashValue: share?.cashAmount ?? .zero,
                disqualificationReasonId: blocked?.reason.rawValue.lowercased(),
                visualColorHex: status == .active ? "#10B981" : "#EF4444"
            )
            let parentId = input.fatherCount > 0 ? "father_1" : rootId
            edges.append(FaraidhGraphEdge(fromId: id, toId: parentId, type: .parentChild))
        }
        
        if input.motherCount > 0 {
            let id = "mother_1"
            let share = activeShares.first { $0.type == .mother }
            nodes[id] = FaraidhGraphNode(
                id: id,
                displayName: "Mother",
                generationLevel: 1,
                relationType: .mother,
                status: share != nil ? .active : .mahjubHirman,
                baseShareFraction: share?.fraction.toDisplayString(),
                finalPercentage: Double(truncating: (share?.percentage ?? .zero) as NSNumber),
                cashValue: share?.cashAmount ?? .zero,
                visualColorHex: share != nil ? "#10B981" : "#EF4444"
            )
            edges.append(FaraidhGraphEdge(fromId: id, toId: rootId, type: .parentChild))
        }
        
        // 3. Children (Generation -1)
        if input.sonCount > 0 {
            for i in 1...input.sonCount {
                let id = "son_\(i)"
                let share = activeShares.first { $0.type == .son }
                let sonPct = Double(truncating: ((share?.percentage ?? .zero) / Decimal(input.sonCount)) as NSNumber)
                let sonCash = (share?.cashAmount ?? .zero) / Decimal(input.sonCount)
                nodes[id] = FaraidhGraphNode(
                    id: id,
                    displayName: "Son \(i)",
                    generationLevel: -1,
                    relationType: .son,
                    status: share != nil ? .active : .mahjubHirman,
                    baseShareFraction: share?.fraction.toDisplayString(),
                    finalPercentage: sonPct,
                    cashValue: sonCash,
                    visualColorHex: share != nil ? "#10B981" : "#EF4444"
                )
                edges.append(FaraidhGraphEdge(fromId: rootId, toId: id, type: .parentChild))
            }
        }
        
        if input.daughterCount > 0 {
            for i in 1...input.daughterCount {
                let id = "daughter_\(i)"
                let share = activeShares.first { $0.type == .daughter }
                let daughterPct = Double(truncating: ((share?.percentage ?? .zero) / Decimal(input.daughterCount)) as NSNumber)
                let daughterCash = (share?.cashAmount ?? .zero) / Decimal(input.daughterCount)
                nodes[id] = FaraidhGraphNode(
                    id: id,
                    displayName: "Daughter \(i)",
                    generationLevel: -1,
                    relationType: .daughter,
                    status: share != nil ? .active : .mahjubHirman,
                    baseShareFraction: share?.fraction.toDisplayString(),
                    finalPercentage: daughterPct,
                    cashValue: daughterCash,
                    visualColorHex: share != nil ? "#10B981" : "#EF4444"
                )
                edges.append(FaraidhGraphEdge(fromId: rootId, toId: id, type: .parentChild))
            }
        }
        
        // 4. Disqualified heirs
        for (idx, dq) in input.disqualifiedHeirs.enumerated() {
            let id = "dq_\(dq.type.rawValue.lowercased())_\(idx)"
            nodes[id] = FaraidhGraphNode(
                id: id,
                displayName: "\(dq.type.rawValue.lowercased().capitalized) (Excluded)",
                generationLevel: (dq.type == .son || dq.type == .daughter) ? -1 : 0,
                relationType: dq.type,
                status: .excludedByLaw,
                disqualificationReasonId: dq.reason.rawValue.lowercased(),
                visualColorHex: "#EF4444"
            )
            edges.append(FaraidhGraphEdge(fromId: rootId, toId: id, type: .parentChild))
        }
        
        return FaraidhFamilyGraph(rootDeceasedId: rootId, nodes: nodes, edges: edges)
    }
}
